### PR TITLE
add job/config e2e-rosa-nightly-sts

### PR DESCRIPTION
**Pull request by yasun1**
Add job/config e2e-rosa-nightly-sts:

1. Add e2e-rosa-nightly-sts job/config in 4.10 period.
2. Separate configuring htpasswd idp from the cluster provision steps.
3. Add rosa-operators-wait-ready to wait the cluster operators to be ready

### New steps

- rosa-idp-htpasswd
- rosa-operators-wait-ready

### CODE REVIEW

File: ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    break
+  fi
+
+  if (( $(date +"%s") - $start_time >= $IDP_TIMEOUT )); then
+    echo "error: Timed out while waiting for the htpasswd idp is ready for login"
+    exit 1
+  fi
+done
+
+# Store kubeconfig
```
Comment (jianzhangbjz) [inline ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh:74]: Based on the  error: `cat: /alabama/.kube/config: No such file or directory`, can we try to print the below info for debugging here?
```yaml
oc whoami 
oc config view
echo ${KUBECONFIG}
```

Comment (jianzhangbjz) [reply ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh:74]: This command will help to locate the KUBECONFIG file.
```yaml
oc --loglevel=8 config view
```

**Comment by yasun1** [comment]
@jianzhangbjz @rioliu-rh Please help review the PR. Although the e2e-rosa-nightly-sts fails again, it fails at the step `openshift-extended-test` in the quoted workflow `openshift-e2e-test-qe`, maybe it is due to some scenarios that can not be run on the rosa cluster. Currently, adding a filter for rosa on the OCP cases is ongoing, in this PR, I use `E2E_RUN_TAGS: '@aws-ipi and @network-openshiftsdn and not @fips'` to test that the whole workflow can be worked as expected.  From now, I think the failure is an expected/allowed failure, it should not block us merge this PR. WDYT?

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml [hunk 2]
```diff
@@ -516,6 +516,17 @@ tests:
     test:
     - ref: openshift-extended-test
     workflow: cucushift-installer-rehearse-gcp-ipi-private
+- as: e2e-rosa-nightly-sts
+  interval: 500h
```
Comment (kuiwang02) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml:520]: @yasun1 suggest not to use interval and change to use cron to control the test exactly.

**Comment by kuiwang02** [comment]
`E2E_RUN_TAGS: '@aws-ipi and @network-openshiftsdn and not @fips'` is not used for golang.
for golang, suggest you add the `TEST_FILTERS: "~ChkUpgrade&;~NonPreRelease&;~Serial&;~Disruptive&;~DisconnectedOnly&;ROSA&"`

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml (hunk 2 above)
Comment (yasun1) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml:520]: make sense. The job is planned to run every 3 weeks(per sprint). I will update it to `0 9 * * */3`.

File: ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+echo "Kubeconfig file: ${KUBECONFIG}"
+cat ${KUBECONFIG} > "${SHARED_DIR}/kubeconfig"
```
Comment (kuiwang02) [inline ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh:76]: @yasun1 per https://github.com/openshift/release/pull/30227#issuecomment-1178563824, maybe need to check co status before exiting this step.
I guess adding idp-htpaswd possible impact the kube-apiserver co.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml (hunk 2 above)
Comment (yasun1) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10__amd64-stable.yaml]: The above cron is not as expected, it means `At 09:00 on every 3rd day-of-week from Wednesday through Sunday`. There is no good way to achieve `run on Wednesday every 3 weeks(per sprint)`, just define it to `0 8 * * 3`, which is `At 08:00 on Wednesday`. Later, let me discuss with the manager about this.

**Comment by yasun1** [comment]
The root cause for the current failure is that there is no '4.10' nightly openshift version on ocm side suddenly, I will research this with the ocm developers. For this PR, let me move it from 4.10 to 4.11 to solve the current issue.

**Comment by yasun1** [comment]
> The root cause for the current failure is that there is no '4.10' nightly openshift version on ocm side suddenly, I will research this with the ocm developers. For this PR, let me move it from 4.10 to 4.11 to solve the current issue.

Currently, we can not create 4.11 account-roles so I need to revert it back to 4.10.

**Comment by yasun1** [comment]
Now the failure `67 failures in cucushift-e2e` is a cases failure, due to we use the old filter to pick up cases, the failure is expected. All the pre/post steps in workflow in the job are executed successfully which also means no issue left for this PR.

**Comment by liangxia** [comment]
Do we have cluster-admin privilege there ?  I can see most of the e2e failures (currently 222 tests failied, with total number of 429, which is too high) are due to lack of admin privilege.

**Comment by yasun1** [comment]
@liangxia Yes, the account given is granted the `cluster-admin` permissions.
`# Grant cluster-admin access to the cluster
rosa grant user cluster-admin --user=${IDP_USER} --cluster=${CLUSTER_ID}`

File: ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+  if (( $(date +"%s") - $start_time >= $IDP_TIMEOUT )); then
+    echo "error: Timed out while waiting for the htpasswd idp to be ready for login"
+    exit 1
+  fi
+done
+
+# Store kubeconfig
+echo "Kubeconfig file: ${KUBECONFIG}"
+cat ${KUBECONFIG} > "${SHARED_DIR}/kubeconfig"
```
Comment (liangxia) [inline ci-operator/step-registry/rosa/idp/htpasswd/rosa-idp-htpasswd-commands.sh:76]: KUBECONFIG is a path to a kubeconfig file.
Use `cat ${KUBECONFIG} > "${SHARED_DIR}/kubeconfig"` is not expected. Perhaps you mean to use `cp` ?

**Comment by liangxia** [comment]
But that kubeconfig is not properly pass through to cucushift e2e, which cause high failures.

**Comment by kuiwang02** [comment]
Per ''/tmp/secret/kubeconfig' -> '/tmp/kubeconfig'' in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/30227/rehearse-30227-periodic-ci-openshift-openshift-tests-private-release-4.10-amd64-stable-e2e-rosa-nightly-sts/1546395986413228032/artifacts/e2e-rosa-nightly-sts/cucushift-e2e/build-log.txt, the kubeconfig should pass through to cucushift-e2e via `${SHARED_DIR}/kubeconfig`.

if the case fails with privilege, I suggest to check if this kubeconfig is really cluster_admin privilege
maybe it is not real cluster_admin privilege with system:admin, so need to choose cases on ROSA.

**Comment by kuiwang02** [comment]
From golang's view, lgtm

**Comment by yasun1** [comment]
@liangxia What I can say is that the `cluster-admin`  on rosa maybe have the different permissions as the one on ocp. This is the value why we need to add the tag `rosa` on the cases to mark they can be run on the rosa cluster.

From the [log ](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/30227/rehearse-30227-periodic-ci-openshift-openshift-tests-private-release-4.10-amd64-stable-e2e-rosa-nightly-sts/1546395986413228032/artifacts/e2e-rosa-nightly-sts/rosa-idp-htpasswd/build-log.txt)of the step `rosa-idp-htpasswd` 

> Login successful.

> You have access to 91 projects, the list has been suppressed. You can list all projects with 'oc projects'

> Using project "default".

This shows the user is a privilege user, otherwise, the regular user does not have the permissions to list 91 projects.

**Comment by yasun1** [comment]
> Per ''/tmp/secret/kubeconfig' -> '/tmp/kubeconfig'' in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/30227/rehearse-30227-periodic-ci-openshift-openshift-tests-private-release-4.10-amd64-stable-e2e-rosa-nightly-sts/1546395986413228032/artifacts/e2e-rosa-nightly-sts/cucushift-e2e/build-log.txt, the kubeconfig should pass through to cucushift-e2e via `${SHARED_DIR}/kubeconfig`.
> 
> if the case fails with privilege, I suggest to check if this kubeconfig is really cluster_admin privilege maybe it is not real cluster_admin privilege with system:admin, so need to choose cases on ROSA.

Again, the user provided here is a `cluster-admin`, not kubeadmin which is the system:admin

**Comment by kuiwang02** [comment]
> > Per ''/tmp/secret/kubeconfig' -> '/tmp/kubeconfig'' in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/30227/rehearse-30227-periodic-ci-openshift-openshift-tests-private-release-4.10-amd64-stable-e2e-rosa-nightly-sts/1546395986413228032/artifacts/e2e-rosa-nightly-sts/cucushift-e2e/build-log.txt, the kubeconfig should pass through to cucushift-e2e via `${SHARED_DIR}/kubeconfig`.
> > if the case fails with privilege, I suggest to check if this kubeconfig is really cluster_admin privilege maybe it is not real cluster_admin privilege with system:admin, so need to choose cases on ROSA.
> 
> Again, the user provided here is a `cluster-admin`, not kubeadmin which is the system:admin

Got it, thanks. It is what I want to say. the kubecnfig pass correctly, but with cluster-admin, not keubeadmin which maybe be required by the case.

**Comment by liangxia** [comment]
Though the kubeconfig provided both has cluster-admin privilege, they have different formats from OCP and ROSA.
```
# kubeconfig from OCP
users:
- name: admin
  user:
    client-certificate-data: ...hiden...
    client-key-data: ...hiden...

# kubeconfig from ROSA
users:
- name: rosa-admin
  user:
    token: ...hiden...
```

**Comment by yasun1** [comment]
@kuiwang02 @liangxia Due to there being no `rosa` or `ROSA` tags added in the cucushift cases and the golang cases, after updating the filters, there is no case picked up in the e2e test, so all the steps are passed, this ci job will be an empty job that only does cluster provision/de-provision.  

Now we need to wait for two tasks:
1. The [OCPQE-11169](https://issues.redhat.com/browse/OCPQE-11169) is finished to support the regular user kubeconfig on cucushift. 
2. The sub-teams add the `@rosa` tag on the cucushift cases and the 'ROSA' tag on the golang cases.

Whether the PR is kept or merged, is just your decision. Whatever it is, it is fine to me.

**Comment by yasun1** [comment]
@kuiwang02 @liangxia Due to there being no `rosa` or `ROSA` tags added in the cucushift cases and the golang cases, after updating the filters, there is no case picked up in the e2e test, so all the steps are passed, this ci job will be an empty job that only does cluster provision/de-provision.  

Now we need to wait for two tasks:
1. The [OCPQE-11169](https://issues.redhat.com/browse/OCPQE-11169) is finished to support the regular user kubeconfig on cucushift. 
2. The sub-teams add the `@rosa` tag on the cucushift cases and the 'ROSA' tag on the golang cases.

Whether the PR is kept or merged, is just your decision. Whatever it is, it is fine to me.

**Comment by kuiwang02** [comment]
> @kuiwang02 @liangxia Due to there being no `@rosa` or `ROSA` tags added in the cucushift cases and the golang cases, after updating the filters, there is no case picked up in the e2e test, so all the steps are passed, this ci job will be an empty job that only does cluster provision/de-provision.
> 
> Now we need to wait for two tasks:
> 
> 1. The [OCPQE-11169](https://issues.redhat.com/browse/OCPQE-11169) is finished to support the regular user kubeconfig on cucushift.
> 2. The sub-teams add the `@rosa` tag on the cucushift cases and the 'ROSA' tag on the golang cases.
> 
> Whether the PR is kept or merged, is just your decision. Whatever it is, it is fine to me.

@yasun1 please use the filter I provided you for golang.
```yaml
TEST_FILTERS: ~ChkUpgrade&;~NonPreRelease&;~Serial&;~Disruptive&;~DisconnectedOnly&;ROSA&
```
not
```yaml
TEST_FILTERS: ~ConnectedOnly&;ROSA&
```
which is not correct.

**Comment by yasun1** [comment]
@kuiwang02  Yes, you are right, my misunderstanding. Let me change the golang's to your suggested.

**Pull request merged**